### DISCUSSION TITLE
Inheritance between normal and simplified objects?

### DISCUSSION THREAD

**Discussion by Selbi182** [spotify-web-api-java/spotify-web-api-java#338] [General]
This isn't a request, so much as it is just a question about how feasible it would be. If the answer is yes, I could see myself making a PR.

So, basically, there are a ton of "simplified" variants of many entities on Spotify. AlbumSimplified, ArtistSimplified, TrackSimplified, PlaylistSimplified, etc. From my own observation, it seems like these are true subsets of their respective "full" versions.

As an example, compare the fields of TrackSimplified to Track:

```
// TrackSimplified
private final ArtistSimplified[] artists;
private final CountryCode[] availableMarkets;
private final Integer discNumber;
private final Integer durationMs;
private final Boolean explicit;
private final ExternalUrl externalUrls;
private final String href;
private final String id;
private final Boolean isPlayable;
private final TrackLink linkedFrom;
private final String name;
private final String previewUrl;
private final Integer trackNumber;
private final ModelObjectType type;
private final String uri;

// Track (full)
private final AlbumSimplified album;
private final ArtistSimplified[] artists;
private final CountryCode[] availableMarkets;
private final Integer discNumber;
private final Integer durationMs;
private final Boolean explicit;
private final ExternalId externalIds;
private final ExternalUrl externalUrls;
private final String href;
private final String id;
private final Boolean isPlayable;
private final TrackLink linkedFrom;
private final Restrictions restrictions;
private final String name;
private final Integer popularity;
private final String previewUrl;
private final Integer trackNumber;
private final ModelObjectType type;
private final String uri;
```

If you toss away every field that is contained in _both_ variants, this is what you're left with:
```
// TrackSimplified
// (no unique fields)

// Track (full)
private final AlbumSimplified album;
private final ExternalId externalIds;
private final Restrictions restrictions;
private final Integer popularity;
```

As you can see, not a whole lot remains unique to the full object and the simplified doesn't have _any_ unique fields.

So, from a design perspective, I'm not quite sure why the full objects don't simply extend the simplified objects. It seems like a ton of duplicated code for little to no benefit. Unless, of course, there is a reason behind it. Maybe it makes JSON parsing easier or something, I don't know.

**Comment by dargmuesli** [comment]
That's true, good observation. If I knew for sure that this will definitely keep being the case, I'd love simplified code! Thing is, I'm not Spotify and I can't be sure. Especially looking at other design choices they made (stuffing tracks and episodes in one field, ...) I'd rather not risk doing a big refactor and having to undo it in the case. Not much is currently changing at the api and the existing code does its job...
When I thought about a rewrite last, I had a Kotlin rewrite in mind as well. We should also switch to Gradle, I'd say. All in all, that would be another rewrite from the ground up for which I currently don't see the benefit for users. It would make for a great school or (small) university project, e.g., and I'd accept the changes, but personally I don't see myself digging into that in the foreseeable future.

**Comment by Selbi182** [comment]
That's fair enough. This would definitely require a massive rewrite in one way or another. 😅
